File: CadenceCore/Sources/CadenceCore/BarbellScene.swift
import Foundation

/// Photographic face colourisation, mirrored by `plateTintMatrix` on web: a
/// 5×4 colour matrix (row-major R, G, B, A rows of five) that rebuilds every
/// channel from the texture's luminance. The plate keeps its photographed
/// shading — rim shadow, rubber grain, machining — and takes its hue from the
/// palette fill, instead of the earlier per-channel gains that clamped the
/// dark textures into one flat colour. The renderer applies it only to the
/// face, then restores the metal hub.
public struct PlateFaceTint: Equatable, Sendable {
    public let matrix: [Double]

    public static let identity: [Double] = [1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0]
    /// Median face luminance of each rendered sprite family (measured on the
    /// shipped PNGs, hub excluded). The lift maps it to 85% of the fill so
    /// the brighter 15% of texels keep headroom before clamping.
    public static func lift(for style: PlateVisualStyle) -> Double {
        0.85 / (style == .bumper ? 0.451 : 0.449)
    }
    /// Fraction of the lift mixed in as grey: highlights whiten instead of
    /// saturating to a single hue.
    public static let greyMix = 0.12

    /// Black iron is the untinted texture; every other token colourises the
    /// face from the palette fill.
    public init(token: String, style: PlateVisualStyle) {
        guard token != "black", let hex = PlatePalette.colours[token]?.fill else {
            matrix = PlateFaceTint.identity
            return
        }
        let fill = [Double((hex >> 16) & 255) / 255, Double((hex >> 8) & 255) / 255, Double(hex & 255) / 255]
        let lift = PlateFaceTint.lift(for: style)
        let grey = lift * PlateFaceTint.greyMix
        let luma = [0.2126, 0.7152, 0.0722]
        var rows: [Double] = []
        for channel in fill {
            let weight = channel * lift + (1 - channel) * grey
            rows += luma.map { $0 * weight } + [0, 0]
        }
        rows += [0, 0, 0, 1, 0]
        matrix = rows
    }

    /// The colour a neutral texel of luminance `l` becomes: (r, g, b), clamped.
    public func apply(luminance l: Double) -> [Double] {
        [0, 5, 10].map { min(1, matrix[$0] * l + matrix[$0 + 1] * l + matrix[$0 + 2] * l) }
    }
}

/// Presentation-only equipment dimensions. These reference profiles do not
/// change a Plate's identity, inventory, recorded mass, or backup encoding.
/// An explicit 5 kg training bumper is full size; a change disc is a different
/// physical profile, never a camera-dependent inference from its weight.
public struct PlateGeometry: Equatable, Sendable {
    public let diameter: Double
    public let thickness: Double

    public init(diameter: Double, thickness: Double) {
        self.diameter = diameter
        self.thickness = thickness
    }

    private static let bumperTable: [String: [Double]] = [
        "25-kg": [450, 70], "20-kg": [450, 60], "15-kg": [450, 48],
        "10-kg": [450, 35], "5-kg": [450, 25],
        "55-lb": [450, 75], "45-lb": [450, 65], "35-lb": [450, 52],
        "25-lb": [450, 40], "10-lb": [450, 25],
    ]
    private static let steelTable: [String: [Double]] = [
        "25-kg": [450, 27], "20-kg": [450, 22], "15-kg": [400, 21],
        "10-kg": [325, 20], "5-kg": [230, 20],
        "55-lb": [450, 30], "45-lb": [450, 27], "35-lb": [400, 25],
        "25-lb": [325, 23], "10-lb": [230, 20],
    ]
    private static let changeTable: [String: [Double]] = [
        "2.5-kg": [210, 19], "2-kg": [190, 19], "1.5-kg": [175, 18],
        "1.25-kg": [160, 16], "1-kg": [160, 16], "0.5-kg": [135, 12],
        "5-lb": [190, 19], "2.5-lb": [160, 16], "1.25-lb": [135, 12],
    ]

    public static func reference(_ plate: Plate, style: PlateVisualStyle) -> PlateGeometry {
        let dimensions = (style == .bumper ? bumperTable : steelTable)[plate.id]
            ?? changeTable[plate.id] ?? [200, 20]
        return PlateGeometry(diameter: dimensions[0], thickness: dimensions[1])
    }

    /// The physical family a denomination belongs to in a style — a full-size
    /// bumper, a calibrated steel disc, or a change plate. It names the cell
    /// in the loadout summary and never changes geometry or mass. Mirrors
    /// web `plateFamily`.
    public static func family(_ plate: Plate, style: PlateVisualStyle) -> String {
        if style == .bumper, bumperTable[plate.id] != nil { return "bumper" }
        if steelTable[plate.id] != nil { return "steel" }
        return "change"
    }

    /// "Bumpers" / "Steel" / "Change" — the cell's second line.
    public static func familyLabel(_ family: String) -> String {
        switch family {
        case "bumper": return "Bumpers"
        case "steel": return "Steel"
        default: return "Change"
        }
    }
}

/// Orthographic scene shared with web barbell-scene.js. Order is the caller's
/// collar-outward order, including deliberately unsorted reverse-mode stacks.
/// Projection affects positions only. Dimensions and mass never change.
public struct BarbellScene: Sendable {
    public struct Disc: Sendable {
        public let plate: Plate
        public let side: Int
        public let index: Int
        public let x: Double
        public let y: Double
        public let radius: Double
        public let faceRadius: Double
        public let depth: Double

        /// The one spoken name for a plate on the bar, on both clients:
        /// "20 kg plate, 1 from inside, left side". Mirrors web
        /// `discAccessibilityLabel`.
        public var accessibilityLabel: String {
            "\(plate.label) plate, \(index + 1) from inside, \(side < 0 ? "left" : "right") side"
        }
    }

    public let discs: [Disc]
    public let width: Double
    public let height: Double
    public let shoulder: Double
    public let end: Double
    public let collar: Double
    public let axisX: Double
    public let axisY: Double
    public let faceScale: Double

    public init(loadout: Loadout, style: PlateVisualStyle, exploded: Bool,
                geometry: [String: PlateGeometry] = [:]) {
        let angle = (exploded ? 38.0 : 18.0) * Double.pi / 180
        axisX = cos(angle)
        axisY = -sin(angle) * 0.24
        faceScale = sin(angle)
        shoulder = loadout.bar == .bar15kg || loadout.bar == .bar35lb ? 145 : 165
        let plates = loadout.perSide.flatMap { Array(repeating: $0.plate, count: max(0, $0.count)) }
        var cursor = shoulder + 8.0
        var previousFaceRadius = 0.0
        var pending: [Disc] = []
        for (index, plate) in plates.enumerated() {
            let shape = geometry[plate.id] ?? PlateGeometry.reference(plate, style: style)
            let radius = max(1, shape.diameter) * 0.18
            let depth = max(1, shape.thickness) * 0.36
            let faceRadius = radius * faceScale
            // Reserve both adjacent projected faces, including a large disc
            // followed by a small change plate. Neither can hide the other.
            if exploded { cursor += (previousFaceRadius + faceRadius + 22) / axisX }
            for side in [-1, 1] {
                let center = Double(side) * (cursor + depth / 2)
                pending.append(Disc(plate: plate, side: side, index: index,
                    x: center * axisX, y: center * axisY, radius: radius,
                    faceRadius: faceRadius, depth: depth * axisX))
            }
            cursor += depth + (exploded ? 0 : 2)
            previousFaceRadius = faceRadius
        }
        collar = cursor + 8
        end = max(shoulder + 150, collar + 28)
        width = max(end * axisX + 25,
            pending.map { abs($0.x) + $0.faceRadius + $0.depth }.max() ?? 0) * 2 + 24
        height = max(110, pending.map { abs($0.y) + $0.radius }.max() ?? 0) * 2 + 40
        // Back-to-front paint order. Stack index remains stable for labels.
        discs = pending.sorted { $0.x < $1.x }
    }
}

File: web/app/js/barbell-scene.js
// Presentation-only reference profiles. Never written into exercise/gym data.
import * as C from "./core.js";
import { plateThemeGeometry } from "./plate-theme.js";

// The one spoken name for a plate on the bar, on both clients. Mirrors
// CadenceCore BarbellScene.Disc.accessibilityLabel.
export const discAccessibilityLabel = (disc) =>
  `${C.plateLabel(disc.plate)} plate, ${disc.index + 1} from inside, ${disc.side < 0 ? "left" : "right"} side`;

const BUMPER = {
  '25-kg': [450, 70], '20-kg': [450, 60], '15-kg': [450, 48], '10-kg': [450, 35], '5-kg': [450, 25],
  '55-lb': [450, 75], '45-lb': [450, 65], '35-lb': [450, 52], '25-lb': [450, 40], '10-lb': [450, 25],
};
const STEEL = {
  '25-kg': [450, 27], '20-kg': [450, 22], '15-kg': [400, 21], '10-kg': [325, 20], '5-kg': [230, 20],
  '55-lb': [450, 30], '45-lb': [450, 27], '35-lb': [400, 25], '25-lb': [325, 23], '10-lb': [230, 20],
};
const CHANGE = {
  '2.5-kg': [210, 19], '2-kg': [190, 19], '1.5-kg': [175, 18], '1.25-kg': [160, 16],
  '1-kg': [160, 16], '0.5-kg': [135, 12], '5-lb': [190, 19], '2.5-lb': [160, 16], '1.25-lb': [135, 12],
};
// The physical family a denomination belongs to in a style. Names the cell in
// the loadout summary; never changes geometry or mass. Mirrors CadenceCore
// PlateGeometry.family / familyLabel.
export function plateFamily(plate, style = 'steel') {
  const key = `${plate.value}-${plate.unit}`;
  if (style === 'bumper' && BUMPER[key]) return 'bumper';
  if (STEEL[key]) return 'steel';
  return 'change';
}
export const plateFamilyLabel = (family) => ({ bumper: 'Bumpers', steel: 'Steel', ipf: 'Steel', machined: 'Steel', iron: 'Iron' })[family] || 'Change';

export function plateGeometry(plate, style = 'steel') {
  const key = `${plate.value}-${plate.unit}`;
  const [diameter, thickness] = (style === 'bumper' ? BUMPER : STEEL)[key] || CHANGE[key] || [200, 20];
  return { diameter, thickness };
}
// Photographic face colourisation, mirrored from CadenceCore PlateFaceTint: a
// 5×4 colour matrix (row-major R, G, B, A rows of five) that rebuilds every
// channel from the texture's luminance, so the plate keeps its photographed
// shading and takes its hue from the palette fill.
export const PLATE_TINT_IDENTITY = [1,0,0,0,0, 0,1,0,0,0, 0,0,1,0,0, 0,0,0,1,0];
// Share of the palette fill each pigment's median face reaches under the scene
// light, matched to the approved plate-loading mockups: coated rubber and
// painted iron read deep and matte (blue deepest), while yellow and white stay
// bright. Black iron keeps its own dark fill.
export const PLATE_TINT_TARGET = Object.freeze({ red: 0.62, blue: 0.49, green: 0.6, yellow: 0.72, white: 0.72, black: 1 });
export const plateTintTarget = (token) => PLATE_TINT_TARGET[token] ?? 0.6;
// Median face luminance of each rendered sprite family (measured, hub excluded).
export const plateTintLift = (style, token) => plateTintLiftForTarget(style, plateTintTarget(token));
const plateTintLiftForTarget = (style, target) => target / (style === "bumper" ? 0.459 : 0.453);
// A theme fill takes the target of the nearest palette pigment (sRGB
// distance): dark iron and black rubber land on black and keep their own fill,
// federation colours on their hue. Mirrors PlateFaceTint.target(forFill:).
export function plateTintTargetForFill(fill) {
  const rgb = (hex) => { const v = Number.parseInt(hex.slice(1), 16); return [16, 8, 0].map((s) => (v >> s) & 255); };
  const a = rgb(fill);
  let best = ["", Infinity];
  for (const token of ["red", "blue", "green", "yellow", "white", "black"]) {
    const b = rgb(C.PLATE_COLOURS[token].fill);
    const d = a.reduce((sum, x, i) => sum + (x - b[i]) * (x - b[i]), 0);
    if (d < best[1]) best = [token, d];
  }
  return plateTintTarget(best[0]);
}
export const PLATE_TINT_GREY_MIX = 0.04;
export function plateTintMatrix(token, style = "steel") {
  const fill = C.PLATE_COLOURS[token]?.fill;
  return fill ? plateTintMatrixForFill(fill, style, plateTintTarget(token)) : PLATE_TINT_IDENTITY;
}
// Colourises the face from any "#rrggbb" fill (a theme's colour rule). target
// overrides the pigment share; null takes the nearest pigment's.
export function plateTintMatrixForFill(fill, style = "steel", target = null) {
  const hex = Number.parseInt(fill.slice(1), 16);
  const channels = [16, 8, 0].map((shift) => ((hex >> shift) & 255) / 255);
  const lift = plateTintLiftForTarget(style, target ?? plateTintTargetForFill(fill)), grey = lift * PLATE_TINT_GREY_MIX;
  const luma = [0.2126, 0.7152, 0.0722];
  const rows = [];
  for (const channel of channels) {
    const weight = channel * lift + (1 - channel) * grey;
    rows.push(...luma.map((l) => l * weight), 0, 0);
  }
  rows.push(0, 0, 0, 1, 0);
  return rows;
}
// The colour a neutral texel of luminance l becomes: [r, g, b], clamped.
export const plateTintApply = (matrix, l) => [0, 5, 10].map((i) => Math.min(1, matrix[i] * l + matrix[i + 1] * l + matrix[i + 2] * l));
export function barbellScene(solution, style = 'steel', exploded = false, geometry = {}, theme = 'custom') {
  const angle = (exploded ? 38 : 18) * Math.PI / 180;
  const axisX = Math.cos(angle), axisY = -Math.sin(angle) * .24, faceScale = Math.sin(angle);
  const shoulder = (solution.bar.unit === 'kg' ? solution.bar.value === 15 : solution.bar.value === 35) ? 145 : 165;
  const plates = solution.perSide.flatMap(c => Array.from({ length: Math.max(0, c.count) }, () => c.plate));
  const discs = [];
  let cursor = shoulder + 8;
  let previousFaceRadius = 0;
  plates.forEach((plate, index) => {
    const shape = geometry[`${plate.value}-${plate.unit}`] || plateThemeGeometry(plate, theme, style);
    const radius = Math.max(1, shape.diameter) * .18, depth = Math.max(1, shape.thickness) * .36;
    const faceRadius = radius * faceScale;
    // Both adjacent faces must fit, especially large plates beside change plates.
    if (exploded) cursor += (previousFaceRadius + faceRadius + 22) / axisX;
    for (const side of [-1, 1]) {
      const center = side * (cursor + depth / 2);
      discs.push({ plate, side, index, x: center * axisX, y: center * axisY,
        radius, faceRadius, depth: depth * axisX, theme });
    }
    cursor += depth + (exploded ? 0 : 2);
    previousFaceRadius = faceRadius;
  });
  const collar = cursor + 8, end = Math.max(shoulder + 150, collar + 28);
  const width = Math.max(end * axisX + 25, ...discs.map(d => Math.abs(d.x) + d.faceRadius + d.depth)) * 2 + 24;
  const height = Math.max(110, ...discs.map(d => Math.abs(d.y) + d.radius)) * 2 + 40;
  discs.sort((a, b) => a.x - b.x);
  return { discs, width, height, shoulder, end, collar, axisX, axisY, faceScale };
}
